alibi_detect/base.py

```diff
@@ -93,40 +93,89 @@ def infer_threshold(self, X: np.ndarray) -> None:
         pass
 
 
+# "Large artefacts" - to save memory these are skipped in _set_config(), but added back in get_config()
+LARGE_ARTEFACTS = ['x_ref', 'c_ref', 'preprocess_fn']
+
+
 class DriftConfigMixin:
     """
-    A mixin class to be used by detector `get_config` methods. The `drift_config` method defines the initial
-    generic configuration dict for all detectors, which is then fully populated by a detector's get_config method(s).
+    A mixin class containing methods related to a drift detector's configuration dictionary.
     """
-    x_ref: np.ndarray
-    preprocess_fn: Optional[Callable] = None
-
-    def drift_config(self):
+    config: Optional[dict] = None
+
+    def get_config(self) -> dict:  # TODO - move to BaseDetector once config save/load implemented for non-drift
+        """
+        Get the detector's configuration dictionary.
+
+        Returns
+        -------
+        The detector's configuration dictionary.
+        """
+        if self.config is not None:
+            # Get config (stored in top-level self)
+            cfg = self.config
+            # Get low-level nested detector (if needed)
+            detector = self._detector if hasattr(self, '_detector') else self  # type: ignore[attr-defined]
+            detector = detector._detector if hasattr(detector, '_detector') else detector  # type: ignore[attr-defined]
+            # Add large artefacts back to config
+            for key in LARGE_ARTEFACTS:
+                if key in cfg:  # self.config is validated, therefore if a key is not in cfg, it isn't valid to insert
+                    cfg[key] = getattr(detector, key)
+            # Set x_ref_preprocessed flag
+            preprocess_at_init = getattr(detector, 'preprocess_at_init', True)  # If no preprocess_at_init, always true!
+            cfg['x_ref_preprocessed'] = preprocess_at_init and detector.preprocess_fn is not None
+            return cfg
+        else:
+            raise NotImplementedError('Getting a config (or saving via a config file) is not yet implemented for this'
+                                      'detector')
+
+    @classmethod
+    def from_config(cls, config: dict):
+        """
+        Instantiate a drift detector from a fully resolved (and validated) config dictionary.
+
+        Parameters
+        ----------
+        config
+            A config dictionary matching the schema's in :class:`~alibi_detect.saving.schemas`.
+        """
+        # Check for exisiting version_warning. meta is pop'd as don't want to pass as arg/kwarg
+        version_warning = config.pop('meta', {}).pop('version_warning', False)
+        # Init detector
+        detector = cls(**config)
+        # Add version_warning
+        detector.meta['version_warning'] = version_warning  # type: ignore[attr-defined]
+        detector.config['meta']['version_warning'] = version_warning
```

Comment:
Not sure if there's much we can do, but noticed that this (and similar patterns) of duplication seem quite common now where we want to keep both `meta` and `config` in sync.

Perhaps for another PR, but what if said `config` is the source of truth and the `meta` attribute just points to the relevant `config` entry (could be a property)?

Reply:
I like that idea, although we'll have to carefully think about how to handle artefacts in the config (`get_config` currently returns a config containing large artefacts, because returning a json-able type config would require the artefacts to be saved, and we had concerns with `get_config` writing to disk). Regardless, definitely worth thinking about what `meta` entails...